src/aatoolbox/datasources/chirps/chirps.py

```diff
@@ -0,0 +1,572 @@
+
+
+class _Chirps(DataSource):
+    """
+    Base class object to retrieve CHIRPS observational precipitation data.
+
+    Parameters
+    ----------
+    country_config : CountryConfig
+        Country configuration
+    geo_bounding_box: GeoBoundingBox
+        the bounding coordinates of the area that should be included in the
+        data.
+    frequency: str
+        time resolution of the data to be downloaded. It can be "daily" or
+        "monthly".
+    resolution: float, default = 0.05
+        resolution of data to be downloaded. Can be 0.05 or 0.25.
+    """
+
+    @abstractmethod
+    def __init__(
+        self,
+        country_config: CountryConfig,
+        geo_bounding_box: GeoBoundingBox,
+        frequency: str,
+        resolution: float = 0.05,
+    ):
+        super().__init__(
+            country_config=country_config,
+            datasource_base_dir="chirps",
+            is_public=True,
+        )
+
+        # round coordinates to correspond with the grid the CHIRPS data is on
+        # non-rounded coordinates can be given to the URL which then
+        # automatically rounds them, but for file saving we prefer to do
+        # this ourselves
+        self._geobb = geo_bounding_box.round_coords(round_val=resolution)
+        self._frequency = frequency
+        self._resolution = resolution
+
+        valid_resolutions = (0.05, 0.25)
+        if resolution not in valid_resolutions:
+            raise ValueError(
+                f"The given resolution is {self._resolution}, which is "
+                "not available. Has to be 0.05 or 0.25."
+            )
+
+    def download(  # type: ignore
+        self,
+        start_date: date = first_available_date,
+        end_date: date = None,
+        clobber: bool = False,
+    ):
+        """
+        Download the CHIRPS observed precipitation as NetCDF file.
+
+        Parameters
+        ----------
+        start_date: datetime.date, default = None
+            Data will be donwloaded starting from date `start_date`.
+            If None, it is set to 1981-1-1.
+        end_date: datetime.date, default = None
+            Data will be donwloaded up to date `end_date`.
+            If None, it is set to the date for which most recent data
+            is available.
+        clobber : bool, default = False
+            If True, overwrites existing raw files
+
+        Returns
+        -------
+        The folder where the data is downloaded.
+        """
+        if end_date is None:
+            end_date = self._get_last_available_date()
+
+        # Create a list of date tuples and a summarising sentence to be printed
+        date_list, sentence_to_print = self._create_date_list(
+            start_date=start_date,
+            end_date=end_date,
+        )
+
+        logger.info(sentence_to_print)
+
+        # Data download
+        for d in date_list:
+            last_filepath = self._download_prep(d=d, clobber=clobber)
+
+        return last_filepath.parents[0]
```

Comment:
Ah that's true. I'm wondering, maybe you can pull `_check_dates_validity` out of `_create_date_list` and have it run before? This way nesting is reduced a bit and it will also be more obvious that indeed the dates were checked.